diff --git a/arch/x86/entry/syscalls/syscall_64.tbl b/arch/x86/entry/syscalls/syscall_64.tbl
index gv6rdgfyg..1d245sdf5e 12453636
--- a/arch/x86/entry/syscalls/syscall_64.tbl
+++ b/arch/x86/entry/syscalls/syscall_64.tbl
@@ -375,6 +375,7 @@
 448	common	landlock_restrict_self	sys_landlock_restrict_self
 449	common	memfd_secret		sys_memfd_secret
 450	common	process_mrelease	sys_process_mrelease
+451	common	kernel_2D_memcpy	sys_kernel_2D_memcpy
 #
 # Due to a historical design error, certain syscalls are numbered differently
 # in x32 as compared to native x86_64.  These syscalls have numbers 512-547.
diff --git a/kernel/sys.c b/kernel/sys.c
index 3564tsdfgh7..scvdxb94 11234564
--- a/kernel/sys.c
+++ b/kernel/sys.c
@@ -422,6 +422,22 @@ SYSCALL_DEFINE2(setregid, gid_t, rgid, gid_t, egid)
 	return __sys_setregid(rgid, egid);
 }
 
+SYSCALL_DEFINE4(kernel_2D_memcpy, float *, Matrix_A, float *, Matrix_B, int, R, int, C)
+{
+	float kernel_matrix[R][C];
+	if(copy_from_user(&kernel_matrix, Matrix_B, sizeof(kernel_matrix)) != 0) 
+	{
+		return -EFAULT;
+	}
+
+	if(copy_to_user(Matrix_A, &kernel_matrix, sizeof(kernel_matrix)) != 0)
+	{
+		return -EFAULT;
+	}	
+
+	return 0;
+}
+
 /*
  * setgid() is implemented like SysV w/ SAVED_IDS
  *
